libraries/AP_SafeRTL/AP_SafeRTL.cpp

```diff
@@ -0,0 +1,584 @@
+    // @Description: SafeRTL maximum number of points on path. Set to 0 to disable SafeRTL.  100 points consumes about 3k of memory.
+    // @Range: 0 500
+    // @User: Advanced
+    // @RebootRequired: True
+    AP_GROUPINFO("POINTS", 1, AP_SafeRTL, _points_max, SAFERTL_POINTS_DEFAULT),
+
+    AP_GROUPEND
+};
+
+/*
+*    This library is used for the Safe Return-to-Launch feature. The vehicle's position
+*    (aka "bread crumbs") are stored into an array in memory at regular intervals.
+*    After a certain number of bread crumbs have been stored and space within the array
+*    is low, clean-up algorithms are run to reduce the total number of points.
+*    When Safe-RTL is initiated by the vehicle code, a more thorough cleanup runs and
+*    the resulting path is fed into navigation controller to return the vehicle to home.
+*
+*    The cleanup consists of two parts, pruning and simplification:
+*
+*    1. Pruning calculates the closest distance between two line segments formed by
+*    two pairs of sequential points, and then cuts out anything between two points
+*    when their line segments get close. This algorithm will never compare two
+*    consecutive line segments. Obviously the segments (p1,p2) and (p2,p3) will
+*    get very close (they touch), but there would be nothing to trim between them.
+*
+*    2. Simplification uses the Ramer-Douglas-Peucker algorithm. See Wikipedia for
+*    a more complete description.
+*
+*    The simplification and pruning algorithms run in the background and do not alter
+*    the path in memory.  Two definitions, SAFERTL_SIMPLIFICATION_TIME_US and
+*    SAFERTL_LOOP_TIME_US are used to limit how long each algorithm will be run
+*    before they save their state and return.
+*
+*    Both algorithm are "anytime algorithms" meaning they can be interrupted before
+*    they complete which is helpful when memory is filling up and we just need to
+*    quickly identify a handful of points which can be deleted.
+*
+*    Once the algorithms have completed the _simplify_complete and _prune_complete
+*    flags are set to true.  The "thorough cleanup" procedure which is run as the
+*    vehicle initiates RTL, waits for these flags to become true.  This can force
+*    the vehicle to pause for a few seconds before initiating the return journey.
+*/
+
+AP_SafeRTL::AP_SafeRTL(const AP_AHRS& ahrs) :
+    _ahrs(ahrs)
+{
+    AP_Param::setup_object_defaults(this, var_info);
+    _simplify_bitmask.setall();
+}
+
+// initialise safe rtl including setting up background processes
+void AP_SafeRTL::init()
+{
+    // protect against repeated call to init
+    if (_path != nullptr) {
+        return;
+    }
+
+    // constrain the path length, in case the user decided to make the path unreasonably long.
+    _points_max = constrain_int16(_points_max, 0, SAFERTL_POINTS_MAX);
+
+    // check if user has disabled SafeRTL
+    if (_points_max == 0 || is_zero(_accuracy)) {
```

Comment:
You should be checking that `!is_positive(accuracy)` instead of zero. (Just to rule out any negative accuracy numbers that might happen by accident).